### PR TITLE
FAst and Reliable haSH

### PR DESCRIPTION
**Pull request by olevino**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
New hash function FARSH

> Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

### CODE REVIEW

File: contrib/farsh-cmake/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,10 @@
+set(LIBRARY_DIR "${ClickHouse_SOURCE_DIR}/contrib/farsh")
```
Comment (alexey-milovidov) [inline contrib/farsh-cmake/CMakeLists.txt:1]: Dynamic dispatch is needed to enable FARSH_AVX2.

File: contrib/farsh-cmake/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+set(SRCS
+        "${LIBRARY_DIR}/farsh.h"
+        "${LIBRARY_DIR}/farsh.c"
+        )
+
+add_library(farsh ${SRCS})
+target_include_directories (farsh PUBLIC ${LIBRARY_DIR})
+
+if(HAVE_AVX2)
```
Comment (alexey-milovidov) [inline contrib/farsh-cmake/CMakeLists.txt:11]: This is wrong.

We should compile multiple versions of the function with different instruction sets and do dynamic dispatch.

I've checked the source code of FARSH and found that it does not have dynamic dispatch.

The code is simple enough, and it is ok to copy-paste it entirely for the purpose of this change.

CC @Bulat-Ziganshin

**Comment by alexey-milovidov** [comment]
It did not make it to production in two months... close?

**Comment by Bulat-Ziganshin** [comment]
FARSH is unfinished code. Please don't use it for any production purposes.

XXHash128 employs the same math, but it's a production-grade hash function.

File: contrib/farsh-cmake/CMakeLists.txt (hunk 2 above)
Comment (yakov-olkhovskiy) [reply contrib/farsh-cmake/CMakeLists.txt:11]: Original code unfortunately is very specific about instruction set, so doesn't fit well. If we really need this hash we can try @Bulat-Ziganshin suggestion:
https://github.com/Cyan4973/xxHash

Comment (alexey-milovidov) [reply contrib/farsh-cmake/CMakeLists.txt:11]: We already use xxHash.

**Pull request closed without merging**